src/gfx/renderables/renderables.ts

```diff
@@ -375,40 +376,41 @@ module Shumway.GFX {
       return renderableBitmap;
     }
 
+    /**
+     * Returns a RenderableBitmap from an Image element, which it uses as its source.
+     *
+     * Takes `width` and `height` as arguments so it can deal with non-decoded images,
+     * which will only get their data after asynchronous decoding has completed.
+     */
+    public static FromImage(image: HTMLImageElement, width: number, height: number) {
+      return new RenderableBitmap(image, new Rectangle(0, 0, width, height));
+    }
+
     public updateFromDataBuffer(type: ImageType, dataBuffer: DataBuffer) {
       if (!imageUpdateOption.value) {
         return;
       }
+      var buffer = dataBuffer.buffer;
       enterTimeline("RenderableBitmap.updateFromDataBuffer", {length: dataBuffer.length});
-      if (type === ImageType.JPEG ||
-          type === ImageType.PNG  ||
-          type === ImageType.GIF)
-      {
-        var self = this;
-        self.setFlags(NodeFlags.Loading);
-        var image = new Image();
-        image.src = URL.createObjectURL(dataBuffer.toBlob(getMIMETypeForImageType(type)));
-        image.onload = function () {
-          self._context.drawImage(image, 0, 0);
-          self.removeFlags(NodeFlags.Loading);
-          self.invalidate();
-        };
-        image.onerror = function () {
-          unexpected("Image loading error: " + ImageType[type]);
-        };
+      if (type === ImageType.JPEG || type === ImageType.PNG || type === ImageType.GIF) {
+        release || Debug.assertUnreachable("Mustn't encounter un-decoded images here");
       } else {
+        var bounds = this._bounds;
+        var imageData = this._imageData;
+        if (!imageData || imageData.width !== bounds.w || imageData.height !== bounds.h) {
+          imageData = new ImageData(new Uint8ClampedArray(dataBuffer.length), bounds.w, bounds.h);
```

Comment:
Well, actually it is different in that fewer temporaries are needed. Still changing it.